File: content/snippets/scala3/design_patterns/visitor.scala
// Modern Scala Visitor using sealed trait ADT + pattern matching (Scala 3)
// No Visitor trait, no accept/visit - shapes are plain ADT cases.
// The compiler enforces exhaustiveness at compile time.

sealed trait Shape
case class Circle()    extends Shape
case class Rectangle() extends Shape

// Operations are plain functions - exhaustive match enforced by the compiler
def draw(shape: Shape): Unit = shape match
  case Circle()    => println("Drawing circle")
  case Rectangle() => println("Drawing rectangle")

// Adding an operation without touching Shape types
def describe(shape: Shape): String = shape match
  case Circle()    => "it's a circle"
  case Rectangle() => "it's a rectangle"

@main def visitorModernExample(): Unit =
  val shapes = List(Circle(), Rectangle())
  shapes.foreach(draw)
  shapes.map(describe).foreach(println)

File: content/snippets/cpp/design_patterns/visitor.cpp
#include <iostream>
#include <memory>

class Visitor;

class Element {
public:
    virtual ~Element() = default;
    virtual void accept(Visitor* visitor) = 0;
};

class Circle : public Element {
public:
    void accept(Visitor* visitor) override;
    void draw() { std::cout << "Drawing circle\n"; }
};

class Rectangle : public Element {
public:
    void accept(Visitor* visitor) override;
    void draw() { std::cout << "Drawing rectangle\n"; }
};

class Visitor {
public:
    virtual ~Visitor() = default;
    virtual void visit(Circle* c) = 0;
    virtual void visit(Rectangle* r) = 0;
};

void Circle::accept(Visitor* visitor) {
    visitor->visit(this);
}

void Rectangle::accept(Visitor* visitor) {
    visitor->visit(this);
}

class DrawVisitor : public Visitor {
public:
    void visit(Circle* c) override {
        c->draw();
    }
    
    void visit(Rectangle* r) override {
        r->draw();
    }
};

int main() {
    auto circle = std::make_shared<Circle>();
    auto rectangle = std::make_shared<Rectangle>();
    auto visitor = std::make_shared<DrawVisitor>();
    
    circle->accept(visitor.get());
    rectangle->accept(visitor.get());
}
